### PR TITLE
Ripsaw class to run various workloads and scale tests

### PR DESCRIPTION
**Pull request by vasukulkarni**
Signed-off-by: Vasu Kulkarni <vasu@redhat.com>

### CODE REVIEW

**Comment by vasukulkarni** [comment]
@tiffanyn108 @wusui - FYI

**Comment by keesturam** [comment]
@vasukulkarni Thank you for sending this PR.

File: ocs_ci/ocs/ripsaw.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+from ocs_ci.ocs.exceptions import CommandFailed
+from ocs_ci.ocs.ocp import OCP
+from ocs_ci.utility.utils import run_cmd
+
+log = logging.getLogger(__name__)
+
+
+class RipSaw(object):
+    """
+      Workload operation using RipSaw
```
Comment (ebenahar) [inline ocs_ci/ocs/ripsaw.py]: align docstring with quotation marks

File: ocs_ci/ocs/ripsaw.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    """
+
+    def __init__(self, **kwargs):
+        """
+        Initializer function
+
+        Args:
+            kwargs (dict):
+                Following kwargs are valid
+                repo: Ripsaw repo to used - a github link
```
Comment (ebenahar) [inline ocs_ci/ocs/ripsaw.py]: please add args types - repo (str): ..

File: ocs_ci/ocs/ripsaw.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            cd_cmd = f'cd {self.dir.name} ; '
+            git_clone_cmd = f'git clone -b {self.branch} {self.repo} '
+            run_cmd(f'{cd_cmd} {git_clone_cmd}')
+            self.crd = 'resources/crd/'
+            self.operator = 'resources/operator.yaml'
+        except CommandFailed as cf:
+            log.error('Error during cloning of ripsaw repository')
+            raise cf
+
+    def apply_crd(self, crd):
```
Comment (ebenahar) [inline ocs_ci/ocs/ripsaw.py]: can you please add a usage example in the docstring? 
Also, don't we need run() and setup() methods as in fio.py?

Comment (vasukulkarni) [reply ocs_ci/ocs/ripsaw.py]: the operator/crd handles it in async way, so we have to add methods to parse the logs properly, those methods will be added when we get better understanding of what to gather and fail on.

File: ocs_ci/ocs/ripsaw.py (hunk 2 above)
Comment (vasukulkarni) [reply ocs_ci/ocs/ripsaw.py]: will fix it.

File: ocs_ci/ocs/ripsaw.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        """
+        Apply the CRD
+
+        Args:
+            crd (str): Name of file to apply
+        """
+        run_cmd(f'cd {self.dir.name} ; oc apply -f deploy')
```
Comment (shylesh) [inline ocs_ci/ocs/ripsaw.py]: We are supposed to cd into ```<ripsawdir>/resources/crds/``` right ?

Comment (vasukulkarni) [reply ocs_ci/ocs/ripsaw.py]: oc command is run from relative path of the git clone.

Comment (vasukulkarni) [reply ocs_ci/ocs/ripsaw.py]: I am yet to push some more changes, Its going to use self.crd and also custom crds.

File: ocs_ci/ocs/ripsaw.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        """
+        self.ocp.new_project(self.namespace)
+
+    def _clone_ripsaw(self):
+        """
+        clone the ripaw repo
+        """
+        self.dir = tempfile.TemporaryDirectory(prefix='ripsaw')
+        try:
+            cd_cmd = f'cd {self.dir.name} ; '
```
Comment (RazTamir) [inline ocs_ci/ocs/ripsaw.py]: Not sure if that was tested already but when I tried to perform git clone, 'cd' operation failed and will need to be executed with:
bash -s 'cd <your_dir>'

File: ocs_ci/ocs/ripsaw.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            git_clone_cmd = f'git clone -b {self.branch} {self.repo} '
+            run_cmd(f'{cd_cmd} {git_clone_cmd}')
```
Comment (vavuthu) [inline ocs_ci/ocs/ripsaw.py]: when run_cmd see ";" its behaving differently ( saying too many arguments ).  we can directly clone to directory using below
```
git_clone_cmd = f'git clone -b {self.branch} {self.repo} {self.dir.name}/ripsaw/'
run_cmd(f'{git_clone_cmd}')
```

File: ocs_ci/ocs/ripsaw.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            log.error('Error during cloning of ripsaw repository')
+            raise cf
+
+    def apply_crd(self, crd):
+        """
+        Apply the CRD
+
+        Args:
+            crd (str): Name of file to apply
+        """
```
Comment (vavuthu) [inline ocs_ci/ocs/ripsaw.py]: Empty line between Args and """
```
Args:
            crd (str): Name of file to apply

"""
```

File: ocs_ci/ocs/ripsaw.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+                repo: Ripsaw repo to used - a github link
+                branch: branch to use from the repo
+                namespace: namespace for the operator
+
+        Example Usage:
+            r1 = RipSaw()
+            r1.apply_crd(crd='ripsaw_v1alpha1_ripsaw_crd.yaml')
+            # use oc apply to apply custom modified bench
+            my_custom_bench = my_custom_bench.yaml
+            run_cmd('oc apply -f my_custom_bench')
```
Comment (vavuthu) [inline ocs_ci/ocs/ripsaw.py]: leave empty line after line 36

File: ocs_ci/ocs/ripsaw.py (hunk 5 above)
Comment (vasukulkarni) [reply ocs_ci/ocs/ripsaw.py]: I have some more changes but this one is working on my mac, I will push my recent changes and attach my run logs.

**Comment by vasukulkarni** [comment]
Still work in progress but logs look like

```
tests/e2e/test_pgsql.py::Test_PgSQL_Workload::test_sql_workload_simple 
--------------------------------------------------------------------------------------------- live log setup ---------------------------------------------------------------------------------------------
15:53:22 - MainThread - tests.conftest - INFO - All logs located at /tmp/ocs-ci-logs-1564613586
15:53:22 - MainThread - tests.conftest - INFO - Running OCS basic installation
15:53:22 - MainThread - tests.conftest - INFO - Openshift Installer will use log level: INFO
15:53:22 - MainThread - ocs_ci.ocs.openshift_ops - INFO - Testing access to cluster with /Users/vasukulkarni/cluster-path/auth/kubeconfig
15:53:22 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc cluster-info
15:53:22 - MainThread - ocs_ci.ocs.openshift_ops - INFO - Access to cluster is OK!
15:53:22 - MainThread - tests.conftest - INFO - The installation is skipped because the cluster is running
15:53:22 - Dummy-1 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get Pod -A -o yaml
15:53:22 - Dummy-2 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get StorageClass -A -o yaml
15:53:22 - Dummy-3 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get CephFileSystem -A -o yaml
15:53:22 - Dummy-4 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get CephBlockPool -A -o yaml
15:53:22 - Dummy-5 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get PersistentVolume -A -o yaml
15:53:22 - Dummy-6 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get PersistentVolumeClaim -A -o yaml
15:53:22 - Dummy-7 - ocs_ci.utility.utils - INFO - Executing command: oc -n rook-ceph --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get Secret  -o yaml
15:53:22 - Dummy-8 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get Namespace -A -o yaml
15:53:32 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc new-project ripsaw
15:53:33 - MainThread - ocs_ci.utility.utils - INFO - Executing command: cd /var/folders/nt/g_hzgqps1q52bc1wbfy3sl_c0000gn/T/ripsawmsfpeonn ;  git clone -b master https://github.com/cloud-bulldozer/ripsaw 
--------------------------------------------------------------------------------------------- live log call ----------------------------------------------------------------------------------------------
15:53:33 - MainThread - ocs_ci.utility.utils - INFO - Executing command: cd /var/folders/nt/g_hzgqps1q52bc1wbfy3sl_c0000gn/T/ripsawmsfpeonn ; oc apply -f deploy
15:53:33 - MainThread - ocs_ci.utility.utils - INFO - Executing command: cd /var/folders/nt/g_hzgqps1q52bc1wbfy3sl_c0000gn/T/ripsawmsfpeonn ; oc apply -f ripsaw_v1alpha1_ripsaw_crd.yaml
15:53:33 - MainThread - ocs_ci.utility.utils - INFO - Executing command: cd /var/folders/nt/g_hzgqps1q52bc1wbfy3sl_c0000gn/T/ripsawmsfpeonn ; oc apply -f resources/operator.yaml
PASSED                                                                                                                                                                                             [100%]
------------------------------------------------------------------------------------------- live log teardown --------------------------------------------------------------------------------------------
15:53:33 - MainThread - ocs_ci.utility.utils - INFO - Executing command: cd /var/folders/nt/g_hzgqps1q52bc1wbfy3sl_c0000gn/T/ripsawmsfpeonn ; oc delete -f ripsaw_v1alpha1_ripsaw_crd.yaml
15:53:33 - MainThread - ocs_ci.utility.utils - INFO - Executing command: cd /var/folders/nt/g_hzgqps1q52bc1wbfy3sl_c0000gn/T/ripsawmsfpeonn ; oc delete -f resources/operator.yaml
15:53:33 - MainThread - ocs_ci.utility.utils - INFO - Executing command: cd /var/folders/nt/g_hzgqps1q52bc1wbfy3sl_c0000gn/T/ripsawmsfpeonn ; oc delete -f deploy
15:53:33 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc delete --wait=true project ripsaw
15:53:34 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get namespace ripsaw -o yaml
15:53:37 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get namespace ripsaw -o yaml
15:53:41 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get namespace ripsaw -o yaml
15:53:41 - MainThread - root - INFO - Error during execution of command: ['oc', '--kubeconfig', '/Users/vasukulkarni/cluster-path/auth/kubeconfig', 'get', 'namespace', 'ripsaw', '-o', 'yaml'].
                                      Error is Error from server (NotFound): namespaces "ripsaw" not found
15:53:41 - Dummy-1 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get Pod -A -o yaml
15:53:41 - Dummy-5 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get StorageClass -A -o yaml
15:53:41 - Dummy-7 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get CephFileSystem -A -o yaml
15:53:41 - Dummy-8 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get CephBlockPool -A -o yaml
15:53:41 - Dummy-3 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get PersistentVolume -A -o yaml
15:53:41 - Dummy-4 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get PersistentVolumeClaim -A -o yaml
15:53:41 - Dummy-2 - ocs_ci.utility.utils - INFO - Executing command: oc -n rook-ceph --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get Secret  -o yaml
15:53:41 - Dummy-6 - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /Users/vasukulkarni/cluster-path/auth/kubeconfig get Namespace -A -o yaml


======================================================================================= 1 passed in 31.07 seconds ========================================================================================
```

**Comment by keesturam** [comment]
@vasukulkarni @tiffanyn108 - Do we plan to use this class for running scale tests? From what I tried with ripsaw manually, it runs a workload for a limited time (~10 minutes) and provides perf numbers. I did not find a way to prolong this test. We can certainly use it for benchmarking, but I'm wondering if we can use it in our regular testing. please let me know.

**Comment by vasukulkarni** [comment]
@keesturam The initial plan is atleast to have the workload running in tier1 (for all supported workloads), and afaik there are many configurable options including run_time eg: https://github.com/cloud-bulldozer/ripsaw/blob/master/docs/pgbench.md,  you can check with Tiffany/Warren, they have been using it for a while. 

In the meantime, lets have some workloads running in our tier tests.

**Comment by keesturam** [comment]
> @keesturam The initial plan is atleast to have the workload running in tier1 (for all supported workloads), 
Sure, makes sense.

File: ocs_ci/ocs/ripsaw.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    def apply_crd(self, crd):
+        """
+        Apply the CRD
+
+        Args:
+            crd (str): Name of file to apply
+        """
+        self.crd = crd
+        self.dir += '/ripsaw'
+        run(f'oc apply -f deploy', shell=True, check=True, cwd=self.dir)
```
Comment (RazTamir) [inline ocs_ci/ocs/ripsaw.py]: Please use exec_oc_command from OCP

File: ocs_ci/ocs/ripsaw.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            )
+            pgsql_cmap = templating.load_yaml_to_dict(
+                constants.PGSQL_CONFIGMAP_YAML
+            )
+            pgsql_sset = templating.load_yaml_to_dict(
+                constants.PGSQL_STATEFULSET_YAML
+            )
+            self.pgsql_service = OCS(**pgsql_service)
+            self.pgsql_service.create()
+            self.pgsql_cmap = OCS(**pgsql_cmap)
```
Comment (RazTamir) [inline ocs_ci/ocs/ripsaw.py]: don't you need to set the kind as weel?

File: ocs_ci/ocs/ripsaw.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+            self.pgsql_cmap.create()
+            self.pgsql_sset = OCS(**pgsql_sset)
```
Comment (RazTamir) [inline ocs_ci/ocs/ripsaw.py]: same

File: ocs_ci/ocs/ripsaw.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+            self.pgsql_sset.create()
+        except CommandFailed as cf:
+            log.error('Failed during setup of PostgreSQL server')
+            raise cf
+        self.pgsql_is_setup = True
+
+    def cleanup(self):
+        run(f'oc delete -f {self.crd}', shell=True, cwd=self.dir)
```
Comment (RazTamir) [inline ocs_ci/ocs/ripsaw.py]: exec_oc_command

File: ocs_ci/ocs/ripsaw.py (hunk 10 above)
Comment (vasukulkarni) [reply ocs_ci/ocs/ripsaw.py]: the kind exists - its coming from yaml (load_to_yaml)

File: ocs_ci/ocs/ripsaw.py (hunk 11 above)
Comment (vasukulkarni) [reply ocs_ci/ocs/ripsaw.py]: same :)

**Comment by tiffanyn108** [comment]
Using this ripsaw.py to deploy ripsaw operator for PGSQL workload, it works well for me.

File: ocs_ci/ocs/ripsaw.py (hunk 9 above)
Comment (vasukulkarni) [reply ocs_ci/ocs/ripsaw.py]: @RazTamir I cannot use that as is without enchanching run_cmd to accept to accept cwd param -> https://github.com/red-hat-storage/ocs-ci/blob/master/ocs_ci/ocs/ocp.py#L70
Here we clone the ripsaw and work from temporary folder and the path has to be relative to that temp folder.

I am going to leave it as is and address later on.

File: ocs_ci/ocs/ripsaw.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+                condition='Running',
+                selector='app=postgres',
+                timeout=120
+            )
+        except (CommandFailed, CalledProcessError) as cf:
+            log.error('Failed during setup of PostgreSQL server')
+            raise cf
+        self.pgsql_is_setup = True
+
+    def cleanup(self):
```
[the remark this answers is not in the document]
Comment (vasukulkarni) [reply ocs_ci/ocs/ripsaw.py]: since the class method is creating the pgsql server, we have to cleanup in our delete.

**Pull request merged**